### PR TITLE
Feature/allow zero numerator

### PR DESCRIPTION
**Pull request by zbintliff**
## Pull Request Checklist

**Is this in reference to an existing issue?**

#### General

- [ ] Update Changelog following the conventions laid out [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

- [x] Update README with any necessary configuration snippets

- [x] Binstubs are created if needed

- [x] RuboCop passes

- [x] Existing tests pass

#### New Plugins

- [ ] Tests

- [ ] Add the plugin to the README

- [ ] Does it have a complete header as outlined [here](http://sensu-plugins.io/docs/developer_guidelines.html#coding-style)

#### Purpose

    Add ability for numerator to be empty

    CloudWatch returns empty data if there is no metric for it. This allows
    us to set a default (0) to replace for the numeratort.

    Remove the --no-data-ok flag because it doesn't make sense with the
    above option. Instead we chang eit to --no-denominator-data-ok. If there
    is no denominator data we will either `unknown` or `ok`.

    Also refactored some to reuse more code. May refactor more.

#### Known Compatibility Issues

Removing/Renaming an CLI option. So this is a breaking change

### CODE REVIEW

**Comment by zbintliff** [comment]
Our current use case is for when we have complete failures and the Number has no data (or is 0). We need to alert on it. Will probably work on tests tomorrow.

**Comment by zbintliff** [comment]
Just wondering if you had any feedback or concerns?

File: lib/sensu-plugins-aws/cloudwatch-common.rb [hunk 1]
```diff
@@ -2,8 +2,17 @@
 module CloudwatchCommon
   include Common
 
-  def client
-    Aws::CloudWatch::Client.new
+  def aws_config
```
Comment (majormoses) [inline lib/sensu-plugins-aws/cloudwatch-common.rb]: why is this needed? Fetching of creds should happen automatically by:
> `include Common`

https://github.com/sensu-plugins/sensu-plugins-aws/blob/10.0.3/lib/sensu-plugins-aws/common.rb#L25-L31

**Comment by majormoses** [comment]
I have not really looked at the code closely but it feels like you are doing exactly what the no data ok option is doing. Can you please expand on how it is different?

**Comment by zbintliff** [comment]
Sorry thought I explained in my commit.  Previously, the `--no-data` option will return ok if numerator or denominator is no data. That's because it was just incorrectly copied from the `check-cloudwatch-metric.rb`.

This allows us to have no data for the numerator, which can be valid. Making the composite 0% so we can actually check it.  Then if then we change the `--no-data` flag to be only for the denominator.

File: lib/sensu-plugins-aws/cloudwatch-common.rb (hunk 1 above)
Comment (zbintliff) [reply lib/sensu-plugins-aws/cloudwatch-common.rb]: I'm not a fan of the way common does it, all the nil checking isn't needed. But I can defer to that.

**Comment by zbintliff** [comment]
Our use case. We are monitoring step functions we want # of activities started/# of activities scheduled.

If its 0/X we care and need to alert. If the denominator is 0 that is fine because it also means the number should also be zero. Previously we have no way of doing that with the check.

File: lib/sensu-plugins-aws/cloudwatch-common.rb (hunk 1 above)
Comment (majormoses) [reply lib/sensu-plugins-aws/cloudwatch-common.rb]: If you think there is a better way to do it in common I am all for it, We just want to have a standard way of fetching aws credentials. Some checks supporting one method but not another leads to a poor user experience and maintenance nightmare.

**Comment by majormoses** [comment]
I'd say we should leave `--no-data-ok` and just add additional options for one or the other as this allows us to keep backwards compatibility and allows people to upgrade without having to change what they are doing. Then its just order of operations checking, if we keep all the existing defaults then we just check the new functions first (where someone specifically wants this functionality) and then if they have not specified anything we fall back to checking as we used to.

**Comment by zbintliff** [comment]
great. i will work on that to keep it backwards compatible.

**Comment by zbintliff** [comment]
Ok @majormoses  so I made a bunch of changes.

1. No longer breaking change (or at least there shouldn't be)
2. Added a bunch of [test](https://github.com/zbintliff/sensu-plugins-aws/blob/feature/allow-zero-numerator/test/lib/cloudwatch_common_spec.rb) for almost the whole `CloudwatchCommon` file.
3. And I feel like i need to defend myself regarding the ugly [if nesting here](https://github.com/zbintliff/sensu-plugins-aws/blob/feature/allow-zero-numerator/lib/sensu-plugins-aws/cloudwatch-common.rb#L65-L92). Basically when you call `ok, warning, critical, or unknown` it automatically exits the script. This is bad when we want to test individual functions. While it works great when using the script, we needed a way to isolate each call to `ok` etc. So we override the automatic exit, which means those functions now only return strings!  So to test each possible branch scenario accurately I had to do it like that. I was going to completely overhaul but didn't want to do too much without feedback.  FWIW though rubo cop didn't complain about block depth (except in one case i fixed)

File: lib/sensu-plugins-aws/common.rb [hunk 2]
```diff
@@ -17,8 +17,8 @@
 #
 
 module Common
-  def initialize
-    super()
+  def initialize(argv = ARGV)
+    super(argv)
```
Comment (zbintliff) [inline lib/sensu-plugins-aws/common.rb]: even though i chose not to do that route, adding this in allows us to spin up a new process and run the script for testing the output and exit code.

File: test/bin/check_ecs_service_health_spec.rb [hunk 3]
```diff
@@ -3,6 +3,9 @@
 require_relative '../ecs_stubs.rb'
 
 describe 'CheckEcsServiceHealth' do
+  after :all do
+    Aws.config = {}
+  end
```
Comment (zbintliff) [inline test/bin/check_ecs_service_health_spec.rb]: this was a bug in this test file that left the AWS Config global object in a bad state. This cleaned it up.

**Comment by majormoses** [comment]
Thanks for this, I gave it a quick look and it looks OK but I want to take a closer look tonight.

**Comment by majormoses** [comment]
#### Regarding nesting conditionals

I mind it a lot less than most ruby folks as I originally hail from python land as a sysadmin. My experiences were forged troubleshooting poorly written applications at 3 AM. As such my concern runs the entire stack and thinks about the poor SOB who is being paged at 3 AM because an app developer wanted to prove how clever they are and missed an obvious edgecase by using an esoteric feature they did not really understand. With the exception of performance readability is always my number one thing I care about. In this case loading the aws sdk is _already expensive_ so in the grand scheme I'd expect that the net result would be negligible. I don't mind seeing code with nested if statements or other very simple and easy to understand logic constructs at least I can probably troubleshoot that in a sleep like state and does not require breaking out the ruby docs to understand the more esoteric features. The same can be said about any language, they all have esoteric features and people use them incorrectly. 

The only reason I dislike this particular example is that the functions that are making the conditionals would have resulted in an exit of the program and in the context of sensu exit codes are very important. I feel like there is a better way of handling this for testing. Would forking the process off to run the test solve this? This allows us to check the status code as well as the output. The integration tests using `serverspec` fork off the command which solves this problem for us with check/metric scripts. I am unsure if we can adapt a similar approach for the unit tests but I think it is at least worth investigating.

#### Lack of Testing in Plugins

I know that testing is something that the plugins generally lack and our lack of testing standards do not help. I am working on changing this and have started pushing serverspec for integration testing. I recently did a couple of posts to make this a little better:
- [general approach and current state](https://blog.sensuapp.org/our-approach-to-testing-sensu-plugins-869260eba94c)
-  [integration testing with plugins](https://blog.sensuapp.org/writing-sensu-plugin-tests-with-test-kitchen-and-serverspec-b646d2eeee51)

Here are some more examples of integration testing using our standard setup:
- https://github.com/sensu-plugins/sensu-plugins-dns/blob/1.3.0/test/integration/helpers/serverspec/check-dns-shared_spec.rb
- https://github.com/sensu-plugins/sensu-plugins-dcos/tree/0.4.0/bin
- https://github.com/sensu-plugins/sensu-plugins-load-checks/tree/4.0.0/test/integration/helpers/serverspec
- https://github.com/sensu-plugins/sensu-plugins-beanstalk/blob/master/test/integration/helpers/serverspec/check-beanstalk_shared_spec.rb

File: lib/sensu-plugins-aws/common.rb (hunk 2 above)
Comment (majormoses) [reply lib/sensu-plugins-aws/common.rb]: I actually just mentioned forking a process for better testing.

**Comment by zbintliff** [comment]
I completely understand the hesitation of the nested ifs.

The biggest reason I didn't fork and run the script is because of testing scope.  If we fork and run we can't isolate which method in `CloudwatchCommon` we are testing. Instead, those forked tests will all fail if one of the methods is bad. Meaning the failed test doesn't indicate what is broken. Currently, every function is mocked and stubbed enough so that only 1 or 2 tests will fail and tied directly to a few lines of code.

With that said I'll get the fork and run stuff up and running when I get a chance.

**Comment by majormoses** [comment]
@eheydrick thoughts?

**Comment by majormoses** [comment]
I guess maybe I feel like library functions should never call `(ok|warning|critical|unknown)` and when things are good/bad they should return what type of result and the result message. This simplifies your problem and has the following benefits:
- it would be a return so the nested ifs go away
- we would then have a way of checking the library functions with unit tests without doing craziness. 
- we would have the ability to more easily write integration tests 

I guess I dislike how much has been pushed into the library, IMHO the scripts are where the alerting logic should go and rely on library functions to address complexity but IMHO they should always just return something they should never exit. This makes it harder to write tests as well as troubleshoot.

**Comment by zbintliff** [comment]
> I guess I dislike how much has been pushed into the library,

Agree 100%, i as just following where it would be. But I'm of the opinion that should be pushed to where the actual check is, since its so specific.

I'd be much happier rewriting it to return the status and string like I did for [check certififcate expiration](https://github.com/sensu-plugins/sensu-plugins-aws/blob/master/bin/check-certificate-expiry.rb#L84) .

I do think we should add integration tests but would rather not add it to this PR as well.

**Comment by majormoses** [comment]
Ya one thing I might add on how to implement it would be to rather than have different flags for crit/warn (unless I too hastily read through it) to just simply return the status soit would look something like this:
##### library
```ruby
def foo()
  # do stuff
  # rescue stuff?
  # return the results in a hash
  {
    status: status,
    message: message
  }
end
```

##### check script
```ruby
def bar()
  # do stuff
  result = foo()
  # exit the script with the appropriate status message and function
  send(result[status], result['message'])
end
```

**Comment by zbintliff** [comment]
Ok @majormoses  updated with the following:
Move `composite_check` into the actual check. Also made it so it return the function to call (via a symbol) and the message.

It is failing rubocop becuase it wants: 
```
      return :ok, "#{metric_desc}: denominator value is zero but that's ok" unless denominator_value.zero? && config[:zero_denominator_data_ok]
      return :unknown, "#{metric_desc}: denominator value is zero" unless denominator_value.zero?
```
instead of: 
```
    if denominator_value.zero? && config[:zero_denominator_data_ok]
      return :ok, "#{metric_desc}: denominator value is zero but that's ok"
    elsif denominator_value.zero?
      return :unknown, "#{metric_desc}: denominator value is zero"
    end
```

Let me know what you prefer.

**Comment by majormoses** [comment]
Cool, I think either is OK I got to run into a couple of meetings but I will review later tonight in more depth. Regarding the `if $CONDITION then return $FOO` vs `return $FOO if $CONDITION` I personally prefer the readability of the conditional then result that being said that comes from my python views and there are actual differences in ruby between them in practicality. I will probably be in and out of meetings for the next few hours so I will try to review and expand on what I said tonight.

**Comment by majormoses** [comment]
One thought about making it more readable, when dealing with guard clauses I do like them when they are short and if you need to check multiple conditionals like in this case you can always make a helper function so you can do something like `p 'true' if var.condition?(var1,var2)`

File: bin/check-cloudwatch-composite-metric.rb [hunk 4]
```diff
@@ -132,7 +132,56 @@ def metric_desc
...
 
+  def composite_check
+    numerator_metric_resp = get_metric(config[:numerator_metric_name])
+    denominator_metric_resp = get_metric(config[:denominator_metric_name])
+
+    ## If the numerator is empty, then we see if there is a default. If there is a default
+    ## then we will pretend the numerator _isnt_ empty. That is
+    ## if empty but there is no default this will be true. If it is empty and there is a default
+    ## this will be false (i.e. there is data, following standard of dealing in the negative here)
+    no_num_data = resp_has_no_data(numerator_metric_resp, config[:statistics]) ? config[:numerator_default].nil? : resp_has_no_data(numerator_metric_resp, config[:statistics])
```
Comment (majormoses) [inline bin/check-cloudwatch-composite-metric.rb]: I had to scroll several times just to read this and I personally prefer readability over shortcuts. Ternary operators are good for some things but usually they are relatively short any longer and I prefer to break them out:
```ruby
no_num_data = if resp_has_no_data(numerator_metric_resp, config[:statistics])
                config[:numerator_default].nil?
              else
                resp_has_no_data(numerator_metric_resp, config[:statistics])
              end
```

File: bin/check-cloudwatch-composite-metric.rb [hunk 5]
```diff
@@ -132,7 +132,56 @@ def metric_desc
...
+    ## Now both the denominator and numerator have data (or a valid default)
+    denominator_value = read_value(denominator_metric_resp, config[:statistics]).to_f
+    if denominator_value.zero? && config[:zero_denominator_data_ok]
+      return :ok, "#{metric_desc}: denominator value is zero but that's ok"
+    elsif denominator_value.zero?
+      return :unknown, "#{metric_desc}: denominator value is zero"
+    end
+
+    ## Only use found metric if there is one else go to the default.
+    numerator_value = resp_has_no_data(numerator_metric_resp, config[:statistics]) ? config[:numerator_default] : read_value(numerator_metric_resp, config[:statistics]).to_f
```
Comment (majormoses) [inline bin/check-cloudwatch-composite-metric.rb]: Same comment regarding ternary operator line length:
```ruby
numerator_value = if resp_has_no_data(numerator_metric_resp, config[:statistics])
                    config[:numerator_default]
                  else
                    read_value(numerator_metric_resp, config[:statistics]).to_f
                  end
```

File: test/bin/check_cloudwatch_composite_metric_spec.rb [hunk 6]
```diff
@@ -31,4 +35,209 @@ def ok(*)
...
+
+  describe '#composite_check' do
+    before :each do
+      Aws.config = { stub_responses: true }
+    end
+
+    after :each do
+      Aws.config = {}
+    end
+    it 'should exit unknown if any data is nil and no flags are passed' do
```
Comment (majormoses) [inline test/bin/check_cloudwatch_composite_metric_spec.rb]: nitpicky but for consistency let's put a space before each assertion

File: test/bin/check_cloudwatch_composite_metric_spec.rb [hunk 7]
```diff
@@ -31,4 +35,209 @@ def ok(*)
...
+      @check.config = config
+      aws_resp = @check.client.stub_data(:get_metric_statistics)
+      allow(@check).to receive(:get_metric).and_return(aws_resp)
+      allow(@check).to receive(:metric_desc).and_return('test')
+
+      resp, msg = @check.composite_check
+      expect(resp).to eq(:unknown)
+      expect(msg).to eq('test could not be retrieved')
+    end
+
```
Comment (majormoses) [inline test/bin/check_cloudwatch_composite_metric_spec.rb]: nitpicky but for consistency let's either put a newline before each assertion (preferable) or kill this line.

File: bin/check-cloudwatch-composite-metric.rb (hunk 4 above)
Comment (majormoses) [reply bin/check-cloudwatch-composite-metric.rb]: Another option would be to do the assignment above with a shorter var name and then use that.

File: test/bin/check_cloudwatch_composite_metric_spec.rb (hunk 7 above)
Comment (zbintliff) [reply test/bin/check_cloudwatch_composite_metric_spec.rb]: Added new line before each assertion

**Comment by zbintliff** [comment]
Broke out a new function that will hopefully increase legibility of finding numerator defaults.

File: test/spec_helper.rb [hunk 8]
```diff
@@ -27,7 +27,7 @@
 #
 module Common
   at_exit do
-    exit! 0
+    @@autorun = false
```
Comment (majormoses) [inline test/spec_helper.rb]: this is no longer needed right?

Comment (zbintliff) [reply test/spec_helper.rb]: its actually an old bug from when i first wrote that spec_helper.rb.  with the `exit! 0` it will "hide" some failed tests. The new line `@@autorun = false` means when you create the new check it won't auto run now.

**Comment by zbintliff** [comment]
The only outstanding issues i can see:
1. Rubocop (gaurd clauses, which i'm not a fan of with the multi return and multi condition)
2. Changelog

**Comment by majormoses** [comment]
> Rubocop (gaurd clauses, which i'm not a fan of with the multi return and multi condition)

Understandable, I think that the principle preached by Dijkstra of single return makes more sense when you have several hundred line functions. Having it return as soon as possible in the application does have some real benefits. Aside from the performance ones it is a great way to short circuit bugs "later" in the application by returning immediately when various conditions are met. You could certainly have a bug causing it to pre-maturely return but that will actually be much easier to troubleshoot as you know where it bailed. I will review the specific example and let you know if I think it can be improved or whether we should disable inline.

> Changelog
:+1:

**Comment by majormoses** [comment]
released: https://rubygems.org/gems/sensu-plugins-aws/versions/10.1.0

**Comment by zbintliff** [comment]
Awesome thanks so much!
